framework/src/wifi_manager/wifi_manager_dhcp_internal.c

```diff
@@ -0,0 +1,209 @@
+wifi_manager_result_e dhcps_start(dhcp_sta_joined cb)
+{
+	struct in_addr ip = {.s_addr = 0x012fa8c0 };
+	struct in_addr netmask = {.s_addr = 0x00ffffff};
+	struct in_addr gw = {.s_addr = 0x012fa8c0};
+
+	_dhcps_remove_list();
+	WIFIMGR_SET_IP4ADDR(WIFIMGR_SOFTAP_IFNAME, ip, netmask, gw);
+
+	if (dhcp_server_start(WIFIMGR_SOFTAP_IFNAME, cb) != OK) {
+		ndbg("[DHCP] DHCP Server - started fail\n");
+		return WIFI_MANAGER_FAIL;
+	}
+	nvdbg("[DHCP] DHCP Server - started successfully\n");
+
+	return WIFI_MANAGER_SUCCESS;
+}
+
+wifi_manager_result_e dhcps_stop(void)
+{
+	struct in_addr in = { .s_addr = INADDR_NONE };
+	WIFIMGR_SET_IP4ADDR(WIFIMGR_SOFTAP_IFNAME, in, in, in);
+
+	if (dhcp_server_stop(WIFIMGR_SOFTAP_IFNAME) != OK) {
+		ndbg("[DHCP] DHCP Server - stopped failed\n");
+	}
+	nvdbg("[DHCP] DHCP Server - stopped successfully\n");
+	return WIFI_MANAGER_SUCCESS;
+}
+
+dhcp_status_e dhcps_add_node(void *msg)
+{
+	dhcp_node_s *node = (dhcp_node_s *)msg;
+	if (node != NULL) {
+		nvdbg("[DHCP] IP: %d:%d:%d:%d\n",
+			  ((char *)&(node->ipaddr))[0],
+			  ((char *)&(node->ipaddr))[1],
+			  ((char *)&(node->ipaddr))[2],
+			  ((char *)&(node->ipaddr))[3]);
+		nvdbg("[DHCP] MAC: %02x:%02x:%02x:%02x:%02x:%02x\n",
+			  node->macaddr[0], node->macaddr[1],
+			  node->macaddr[2], node->macaddr[3],
+			  node->macaddr[4], node->macaddr[5]);
+	} else {
+		ndbg("[DHCP] No DHCP node exists\n");
+		return DHCP_ERR;
+	}
+
+	if (memcmp(g_dhcp_list.macaddr, node->macaddr, 6) == 0) {
+		g_dhcp_list.ipaddr = node->ipaddr;
+		return DHCP_EXIST;
+	}
+	g_dhcp_list.ipaddr = node->ipaddr;
+	memcpy(g_dhcp_list.macaddr, node->macaddr, 6);
+	return DHCP_OK;
+}
+
+void dhcps_del_node(void)
+{
+	//TODO: Currently only one client can be joined and left
+	g_dhcp_list.ipaddr = 0;
+	for (int i = 0; i < 6; i++) {
+		g_dhcp_list.macaddr[i] = 0;
+	}
+}
+
+static void _dhcps_remove_list(void)
+{
+	//TODO: Currently only one client can be joined and left
+	g_dhcp_list.ipaddr = 0;
+	for (int i = 0; i < 6; i++) {
+		g_dhcp_list.macaddr[i] = 0;
+	}
+}
+#endif //CONFIG_WIFIMGR_DISABLE_DHCPS
+
+
+/**
+ * Internal DHCP client APIs
+ */
+#ifndef CONFIG_WIFIMGR_DISABLE_DHCPC
+wifi_manager_result_e dhcpc_get_ipaddr(void)
+{
+	int ret;
+	struct in_addr ip;
+	wifi_manager_result_e wret = WIFI_MANAGER_FAIL;
+
+	ret = dhcp_client_start(WIFIMGR_STA_IFNAME);
+	if (ret != OK) {
+		WIFIADD_ERR_RECORD(ERR_WIFIMGR_CONNECT_DHCPC_FAIL);
+		ndbg("[DHCPC] get IP address fail\n");
+		return wret;
+	}
+
+	ret = netlib_get_ipv4addr(WIFIMGR_STA_IFNAME, &ip);
+	if (ret != OK) {
+		ndbg("[DHCPC] get IP address fail\n");
+		WIFIADD_ERR_RECORD(ERR_WIFIMGR_CONNECT_DHCPC_FAIL);
+		return wret;
+	}
+	nvdbg("[DHCPC] get IP address %s\n", inet_ntoa(ip));
+
+	return WIFI_MANAGER_SUCCESS;
+}
+#endif //CONFIG_WIFIMGR_DISABLE_DHCPC
+
+//Set IP to zero, regardless of DHCPC status
+void dhcpc_close_ipaddr(void)
+{
+#ifndef CONFIG_WIFIMGR_DISABLE_DHCPC
+	dhcp_client_stop(WIFIMGR_STA_IFNAME);
```

Comment:
long time ago, I gave comment to someone, dhcp_client_stop calls socket & ioctl function but type is void.
it must check return value.(change type to none-void is needed).
Additionally, this is framework, so you have to make some routine to handle it when it failed(For example, reset every thing forcely or retry)

Reply:
DHCP implemented in LwIP has no return for closing, and we have decided to follow the concept from it.  Moreover, if this function fails, then LwIP (inside ioctl) will make an assert because it is very critical, which might not require a recovery logic here, only for this API.

Thanks for the comment.

Reply:
Why you think only ioctl? socket & close also need to check.
when you expand this code more, working something with invalid socket will bring many problems.
Also if someone use this wrongly, it will make memory leak continuously, because you didn't any error message after close.
It is very critical, you should consider this very carefully. 
If I were you I would fix & contribute to lwip project.

Reply:
Actually, there are two definitions of **dhcp_client_stop**, depending on the initial configuration in menuconfig. One of them comes from external (uIP-based) DHCPC and the other is LwIP DHCPC.
The former simply sets the IP address of the target network interface to the very initial value directly, and the latter conducts the normal socket/ioctl operation as usual. That means, socket fd is verified, and then ioctl is called, and if a function for the specific ioctl fails then, LwIP will make an assert. 
Please let me know what should I check for better implementation.